nvme_gw_server.py

```diff
@@ -196,15 +199,65 @@ def start_spdk(self):
         except Exception as ex:
             self.logger.error(f"Unable to initialize SPDK: \n {ex}")
             raise
-        return
+
+    def create_transport(self, trtype):
+        if trtype in self.transports:
+            self.logger.debug(
+                f"create_transport: transport {trtype} already exists"
+            )
+            return
+
+        args = {'trtype' : trtype}
+        name = "transport_" + trtype + "_options"
+
+        options = self.nvme_config.get("spdk", name, "")
+
+        SUPPORTED_OPTIONS = {
+            'max_queue_depth' : int,
+            'max_io_qpairs_per_ctrlr' : int,
+            'in_capsule_data_size' : int,
+            'max_io_size' : int,
+            'io_unit_size' : int,
+            'max_aq_depth' : int,
+            'num_shared_buffers' : int,
+            'buf_cache_size' : int,
+            'dif_insert_or_strip' : lambda x : bool(strtobool(x)),
+            'abort_timeout_sec' : int,
+            'zcopy' : lambda x : bool(strtobool(x)),
+            'tgt_name' : str,
+            'acceptor_poll_rate' : int,
+        }
+
+        self.logger.debug(f"create_transport: {trtype} options: {options}")
+
+        for opt in options.split():
+            try:
+                k, v =opt.split('=')
+                if k in SUPPORTED_OPTIONS:
+                    args[k] = SUPPORTED_OPTIONS[k](v)
+                else:
+                    self.logger.error(f"Ignoring unknown {name} option {k}={v}")
```

Comment:
Do we really want to ignore? This leads to the case where you think your option was set but instead it was just ignored. I would fail if a wrong option is specified.

Reply:
You will see an error message about this, so it would not be completely ignored.

Actually, I don't have a strong opinion about this. In my initial version, when it parsed on the gateway start I made it fail (and the gateway terminate), because it would indeed make the error detection easier.

But when the gateway is already running and is expected to work, and then a user requests fails due to a wrong thing in the config, it might be confusing for a user to see its "add listener" request to fail.

Anyway, as I said, I don't have a strong opinion, I may change it return the error if you still prefer this way.

Reply:
@sskaur what do you think?

Reply:
Maybe we can parse all the existing options on startup and fail if they are unsupported, so we don't have the issue of failing later here.

Reply:
If parsing on startup, we would need to try all possible `transport_{trtype}_options`, even if one is not going to be used. Then do we want to store the parsing result to use when needed or parse it again?

So right now we have 3 options:

1) the current implementation
2) to fail the `create_listener` request if parsing the transport options failed
3) parse the transport options on startup.

Listed in my preference order but actually I am fine with any, just lets decide which one.

Reply:
As I look at this, a fourth option might be to require that `transport_{trtype}_options` is specified in the config file as a JSON string or in some other self-describing format and avoid parsing altogether.  Just convert the string to a dict and immediately unpack.  If `nvmf_create_transport` call fails, `create_create_listener` fails too.  That way we don't have to keep the list of known options in sync with SPDK when bumping the submodule either.

Reply:
Interesting. It indeed would easy maintaining for us, though offloading it to users. And comparing a possible configuration:
```
transport_tcp_options = max_queue_depth=16 max_io_size=4194304 io_unit_size=1048576 zcopy=false
transport_tcp_options = {"max_queue_depth" : 16, "max_io_size" : 4194304, "io_unit_size" : 1048576, "zcopy" : false}
```
I suppose the first way would be a little easier for users (I had to test the json string in python to make sure it worked, and ofc it did not work for the first try) but this is probably not a big deal.

So I am fine with it. The main issue to me with this solution though is that it will only fail when a user calls `create_create_listener`, not on startup. And taking that a person that will call `create_create_listener` may be not the same who created the configuration, it will be very difficult for the caller to understand what is wrong with her `create_create_listener` request.

Reply:
> Interesting. It indeed would easy maintaining for us, though offloading it to users. And comparing a possible configuration:
> 
> ```
> transport_tcp_options = max_queue_depth=16 max_io_size=4194304 io_unit_size=1048576 zcopy=false
> transport_tcp_options = {"max_queue_depth" : 16, "max_io_size" : 4194304, "io_unit_size" : 1048576, "zcopy" : false}
> ```
> 
> I suppose the first way would be a little easier for users (I had to test the json string in python to make sure it worked, and ofc it did not work for the first try) but this is probably not a big deal.

I realize that and to that end I suggested switching the config file format to YAML while it's not too late.

> 
> So I am fine with it. The main issue to me with this solution though is that it will only fail when a user calls `create_create_listener`, not on startup. And taking that a person that will call `create_create_listener` may be not the same who created the configuration, it will be very difficult for the caller to understand what is wrong with her `create_create_listener` request.

It's funny how you picked up my `create_create_listener` typo ;)

Yeah, I was reacting the redundant parsing aspect mostly.  Given that a transport of a particular type is a global entity, creating it on startup and potentially failing there makes sense to me.  I'm going to leave this aspect (completely orthogonal to redundant parsing) to @PepperJo who suggested the move to `create_listener`.

Reply:
Makes sense to me. Let's keep the parsing to SPDK, so we don't have to keep track of all allowed options. We can move it back to the startup, the only downside is that we might create a transport that is never used, but it solves the problem with failing when creating the listener.

Reply:
Just the ones in the config like in your initial version.